### ISSUE TITLE
ValueError when handling event

### ISSUE THREAD

**Issue by ghcs27** [SoCo/SoCo#458]
I have an avTransport subscription from a Sonos speaker, and every now and again I get a value error:
```
Exception happened during processing of request from ('192.168.123.16', 4911)
Traceback (most recent call last):
  File "usr/lib/python3.4/socketserver.py", line 613, in process_request_thread
    self.finish_request(request, client_adress)
  File "usr/lib/python3.4/socketserver.py", line 344, in finish_request
    self.RequestHandlerClass(request, client_adress, self)
  File "usr/lib/python3.4/http/server.py", line 398, in handle
    self.handle_one_request()
  File "usr/lib/python3.4/http/server.py", line 386, in handle_one_request
    method()
  File "usr/local/lib/python3.4/dist-packages/soco/events.py", line 227, in do_NOTIFY
    variables = parse_event_xml(content)
  File "usr/local/lib/python3.4/dist-packages/soco/events.py", line 130, in parse_event_xml
    value = from_didl_string(value)[0]
  File "usr/local/lib/python3.4/dist-packages/soco/data_structures.py", line 93, in from_didl_string
    items.append(cls.from_element(elt))
  File "usr/local/lib/python3.4/dist-packages/soco/data_structures.py", line 501, in from_element
    title = really_unicode(title_elt.text)
  File "usr/local/lib/python3.4/dist-packages/soco/utils.py", line 43, in really_unicode
    raise ValueError('%s is not a string at all.' % in_string)
ValueError: None is not a string at all.
```
It seems like in line 501 of data_structures.py the title tag is present in the XML, but has no content. Maybe title_elt.text should be checked for None values and replaced with an empty string.
Would someone please take a look at this?

I am using soco 0.11.1 on Python 3.4.

**Comment by ghcs27** [comment]
Has anyone else seen this error? I am really desperate because now every event triggers the exception so I cannot receive any events.

**Comment by DPH** [comment]
Hi,
I use avTransport subscription and can handle them ok. 
I do have try...except blocks in my code, and test for blanks, but it works. 
What are you trying to do? Can you post your code?

Cheers Daivd

**Comment by ghcs27** [comment]
The problem seems to be at least partly due to what the player is sending. I've tried just executing
```
player = SoCo('192.168.123.16')
sub = player.avTransport.subscribe(auto_renew=True)
```
and I start getting an exception whenever an event occurs. When I do `sub.events.get()` the queue is still empty. I will try to catch the XML the player sends - maybe that will bring some light into the dark.

**Comment by ghcs27** [comment]
The event content is
```
<e:propertyset xmlns:e="urn:schemas-upnp-org:event-1-0">
 <e:property>
  <LastChange>
   &lt;Event xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/AVT/&quot; xmlns:r=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot;&gt;&lt;InstanceID val=&quot;0&quot;&gt;&lt;TransportState val=&quot;PAUSED_PLAYBACK&quot;/&gt;&lt;CurrentPlayMode val=&quot;NORMAL&quot;/&gt;&lt;CurrentCrossfadeMode val=&quot;0&quot;/&gt;&lt;NumberOfTracks val=&quot;298&quot;/&gt;&lt;CurrentTrack val=&quot;1&quot;/&gt;&lt;CurrentSection val=&quot;0&quot;/&gt;&lt;CurrentTrackURI val=&quot;x-file-cifs://digitus/simone/Adele/Adele%20-%2021%20(2011)/01%20Rolling%20in%20the%20Deep.mp3&quot;/&gt;&lt;CurrentTrackDuration val=&quot;0:03:49&quot;/&gt;&lt;CurrentTrackMetaData val=&quot;&amp;lt;DIDL-Lite xmlns:dc=&amp;quot;http://purl.org/dc/elements/1.1/&amp;quot; xmlns:upnp=&amp;quot;urn:schemas-upnp-org:metadata-1-0/upnp/&amp;quot; xmlns:r=&amp;quot;urn:schemas-rinconnetworks-com:metadata-1-0/&amp;quot; xmlns=&amp;quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&amp;quot;&amp;gt;&amp;lt;item id=&amp;quot;-1&amp;quot; parentID=&amp;quot;-1&amp;quot; restricted=&amp;quot;true&amp;quot;&amp;gt;&amp;lt;res protocolInfo=&amp;quot;x-file-cifs:*:audio/mpeg:*&amp;quot; duration=&amp;quot;0:03:49&amp;quot;&amp;gt;x-file-cifs://digitus/simone/Adele/Adele%20-%2021%20(2011)/01%20Rolling%20in%20the%20Deep.mp3&amp;lt;/res&amp;gt;&amp;lt;r:streamContent&amp;gt;&amp;lt;/r:streamContent&amp;gt;&amp;lt;r:radioShowMd&amp;gt;&amp;lt;/r:radioShowMd&amp;gt;&amp;lt;upnp:albumArtURI&amp;gt;/getaa?u=x-file-cifs%3a%2f%2fdigitus%2fsimone%2fAdele%2fAdele%2520-%252021%2520(2011)%2f01%2520Rolling%2520in%2520the%2520Deep.mp3&amp;amp;amp;v=141&amp;lt;/upnp:albumArtURI&amp;gt;&amp;lt;dc:title&amp;gt;Rolling in the Deep&amp;lt;/dc:title&amp;gt;&amp;lt;upnp:class&amp;gt;object.item.audioItem.musicTrack&amp;lt;/upnp:class&amp;gt;&amp;lt;dc:creator&amp;gt;Adele&amp;lt;/dc:creator&amp;gt;&amp;lt;upnp:album&amp;gt;21&amp;lt;/upnp:album&amp;gt;&amp;lt;upnp:originalTrackNumber&amp;gt;1&amp;lt;/upnp:originalTrackNumber&amp;gt;&amp;lt;r:albumArtist&amp;gt;Adele&amp;lt;/r:albumArtist&amp;gt;&amp;lt;/item&amp;gt;&amp;lt;/DIDL-Lite&amp;gt;&quot;/&gt;&lt;r:NextTrackURI val=&quot;x-file-cifs://digitus/simone/Adele/Adele%20-%2021%20(2011)/02%20Rumour%20Has%20It.mp3&quot;/&gt;&lt;r:NextTrackMetaData val=&quot;&amp;lt;DIDL-Lite xmlns:dc=&amp;quot;http://purl.org/dc/elements/1.1/&amp;quot; xmlns:upnp=&amp;quot;urn:schemas-upnp-org:metadata-1-0/upnp/&amp;quot; xmlns:r=&amp;quot;urn:schemas-rinconnetworks-com:metadata-1-0/&amp;quot; xmlns=&amp;quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&amp;quot;&amp;gt;&amp;lt;item id=&amp;quot;-1&amp;quot; parentID=&amp;quot;-1&amp;quot; restricted=&amp;quot;true&amp;quot;&amp;gt;&amp;lt;res protocolInfo=&amp;quot;x-file-cifs:*:audio/mpeg:*&amp;quot; duration=&amp;quot;0:03:43&amp;quot;&amp;gt;x-file-cifs://digitus/simone/Adele/Adele%20-%2021%20(2011)/02%20Rumour%20Has%20It.mp3&amp;lt;/res&amp;gt;&amp;lt;upnp:albumArtURI&amp;gt;/getaa?u=x-file-cifs%3a%2f%2fdigitus%2fsimone%2fAdele%2fAdele%2520-%252021%2520(2011)%2f02%2520Rumour%2520Has%2520It.mp3&amp;amp;amp;v=141&amp;lt;/upnp:albumArtURI&amp;gt;&amp;lt;dc:title&amp;gt;Rumour Has It&amp;lt;/dc:title&amp;gt;&amp;lt;upnp:class&amp;gt;object.item.audioItem.musicTrack&amp;lt;/upnp:class&amp;gt;&amp;lt;dc:creator&amp;gt;Adele&amp;lt;/dc:creator&amp;gt;&amp;lt;upnp:album&amp;gt;21&amp;lt;/upnp:album&amp;gt;&amp;lt;upnp:originalTrackNumber&amp;gt;2&amp;lt;/upnp:originalTrackNumber&amp;gt;&amp;lt;r:albumArtist&amp;gt;Adele&amp;lt;/r:albumArtist&amp;gt;&amp;lt;/item&amp;gt;&amp;lt;/DIDL-Lite&amp;gt;&quot;/&gt;&lt;r:EnqueuedTransportURI val=&quot;file:///jffs/settings/savedqueues.rsq#7&quot;/&gt;&lt;r:EnqueuedTransportURIMetaData val=&quot;&amp;lt;DIDL-Lite xmlns=&amp;quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&amp;quot; xmlns:dc=&amp;quot;http://purl.org/dc/elements/1.1/&amp;quot; xmlns:upnp=&amp;quot;urn:schemas-upnp-org:metadata-1-0/upnp/&amp;quot;&amp;gt;&amp;lt;item id=&amp;quot;&amp;quot; parentID=&amp;quot;&amp;quot; restricted=&amp;quot;true&amp;quot;&amp;gt;&amp;lt;dc:title /&amp;gt;&amp;lt;res protocolInfo=&amp;quot;x-rincon-playlist:*:*:*&amp;quot;&amp;gt;file:///jffs/settings/savedqueues.rsq#7&amp;lt;/res&amp;gt;&amp;lt;upnp:class&amp;gt;object&amp;lt;/upnp:class&amp;gt;&amp;lt;desc id=&amp;quot;cdudn&amp;quot; nameSpace=&amp;quot;urn:schemas-rinconnetworks-com:metadata-1-0/&amp;quot;&amp;gt;RINCON_AssociatedZPUDN&amp;lt;/desc&amp;gt;&amp;lt;/item&amp;gt;&amp;lt;/DIDL-Lite&amp;gt;&quot;/&gt;&lt;PlaybackStorageMedium val=&quot;NETWORK&quot;/&gt;&lt;AVTransportURI val=&quot;x-rincon-queue:RINCON_000E5855B28401400#0&quot;/&gt;&lt;AVTransportURIMetaData val=&quot;&quot;/&gt;&lt;NextAVTransportURI val=&quot;&quot;/&gt;&lt;NextAVTransportURIMetaData val=&quot;&quot;/&gt;&lt;CurrentTransportActions val=&quot;Set, Play, Stop, Pause, Seek, Next, Previous&quot;/&gt;&lt;r:CurrentValidPlayModes val=&quot;SHUFFLE, REPEAT, CROSSFADE&quot;/&gt;&lt;TransportStatus val=&quot;OK&quot;/&gt;&lt;r:SleepTimerGeneration val=&quot;0&quot;/&gt;&lt;r:AlarmRunning val=&quot;0&quot;/&gt;&lt;r:SnoozeRunning val=&quot;0&quot;/&gt;&lt;r:RestartPending val=&quot;0&quot;/&gt;&lt;TransportPlaySpeed val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;CurrentMediaDuration val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;RecordStorageMedium val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;PossiblePlaybackStorageMedia val=&quot;NONE, NETWORK&quot;/&gt;&lt;PossibleRecordStorageMedia val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;RecordMediumWriteStatus val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;CurrentRecordQualityMode val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;PossibleRecordQualityModes val=&quot;NOT_IMPLEMENTED&quot;/&gt;&lt;/InstanceID&gt;&lt;/Event&gt;
  </LastChange>
 </e:property>
</e:propertyset>
```

**Comment by ghcs27** [comment]
Possible solutions range from not parsing the EnqueuedTransportURIMetaData (not very user-friendly) over dealing with empty title tags in from_element() in data_structures.py to accepting None values in really_unicode() (convert them to an empty string).
What would you suggest? I will make a bugfix PR as soon as possible.

**Comment by DPH** [comment]
Hi,
the event content above does have a title:
```
<DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/">
	<item id="-1" parentID="-1" restricted="true">
		<res protocolInfo="x-file-cifs:*:audio/mpeg:*" duration="0:03:49">x-file-cifs://digitus/simone/Adele/Adele%20-%2021%20(2011)/01%20Rolling%20in%20the%20Deep.mp3</res>
		<r:streamContent/>
		<r:radioShowMd/>
		<upnp:albumArtURI>/getaa?u=x-file-cifs%3a%2f%2fdigitus%2fsimone%2fAdele%2fAdele%2520-%252021%2520(2011)%2f01%2520Rolling%2520in%2520the%2520Deep.mp3&amp;v=141</upnp:albumArtURI>
		<dc:title>Rolling in the Deep</dc:title>
		<upnp:class>object.item.audioItem.musicTrack</upnp:class>
		<dc:creator>Adele</dc:creator>
		<upnp:album>21</upnp:album>
		<upnp:originalTrackNumber>1</upnp:originalTrackNumber>
		<r:albumArtist>Adele</r:albumArtist>
	</item>
</DIDL-Lite>
```
I see you are using Python 3.4, which SoCo version are you using?
Can I suggest you pull down the latest from https://github.com/SoCo/SoCo
The line numbers that error do not appear to relate to the latest code. 

I use SoCo to watch what is playing on one player and display the details on a screen and it appears to cope with any situation. 

Try running this example - taken from the example in events.py:
```
from queue import Empty
import soco
from pprint import pprint
from soco.events import event_listener

player = soco.SoCo('192.168.1.68')
print(player.player_name)

sub = player.avTransport.subscribe()

while True:
    try:
        event = sub.events.get(timeout=0.5)
        pprint(event.variables)

        curr_track_meta_data = event.current_track_meta_data
        if curr_track_meta_data:
            try:
                title = curr_track_meta_data.title
            except AttributeError:
                title = ''
            print('Track title is:', title)
    except Empty:
        pass

    except KeyboardInterrupt:
        sub.unsubscribe()
        event_listener.stop()
        break
```
Hope this helps
Let us know if it does or you need more help
Cheers David

**Comment by ghcs27** [comment]
When I run your example, I still get exceptions and no events are printed out. I am using the latest version of SoCo.
I just realized that this error doesn't occur after a player restart and reappears when I call add_uri_to_queue(). It seems like the problem is known - see #448.
I might rewrite my code to use add_to_queue() with a proper title instead, but I still feel like SoCo should be able to handle an empty title tag for the EnqueuedTransportURIMetaData.

**Comment by DPH** [comment]
Agree, it should not fail on no title, although not a frequent occurrence SoCo should handle it. 
If you have some ideas on how to fix please raise a pull request. 
I had a quick look but am not sure how to handle this exception.

Cheers David

**Comment by ghcs27** [comment]
There is a fix PR at #463 now. Please have a look at it.

**Comment by KennethNielsen** [comment]
Nice work hammering out the problem here. I'm currently reviewing #467 /by @ghcs27) which should fix this.
